Perform re-sync periodically

johannaengland opened the pull request:
Closes #17. 

The only thing I am uncertain about if a re-sync should be done the first time this function is entered.

johannaengland committed:
Perform re-sync periodically

To avoid Argus and NAV getting out of sync and having to manually trigger re-sync

lunkwill42 reviewed (changes requested):
Nice :)

> The only thing I am uncertain about if a re-sync should be done the first time this function is entered.

Well, I guess that's the thing about this: It changes the semantics of the "sync-on-startup" option.  If left enabled, I guess this change will in fact cause **two** sync operations in sequence before starting to listen for incoming events.

Perhaps the sync-on-startup option should be made completely obsolete by this.  I think I would rather see a "resync interval" setting.  If no interval is set, only the `--sync` option would force a sync.  If an interval is set, your code should operate exactly as it already does, just that the interval comes from a config option rather than a constant.

johannaengland commented:
> Nice :)
> 
> > The only thing I am uncertain about if a re-sync should be done the first time this function is entered.
> 
> Well, I guess that's the thing about this: It changes the semantics of the "sync-on-startup" option. If left enabled, I guess this change will in fact cause **two** sync operations in sequence before starting to listen for incoming events.
> 
> Perhaps the sync-on-startup option should be made completely obsolete by this. I think I would rather see a "resync interval" setting. If no interval is set, only the `--sync` option would force a sync. If an interval is set, your code should operate exactly as it already does, just that the interval comes from a config option rather than a constant.

Oh, I really like this! But I would argue that we should have a default for syncing (maybe one minute) and if someone does not want re-syncing, they should explicitly set that setting to None

johannaengland committed:
Replace sync on startup with sync-interval setting

This way the user can set how often a re-sync is run
If it is set to null no re-sync will be run

lunkwill42 commented on src/navargus/glue.py:
```diff
@@ -567,9 +568,15 @@ def get_api_timeout(self) -> float:
-    def get_sync_on_startup(self):
-        """Returns True if this program should always sync the Argus API on startup"""
-        return bool(self.get("api", {}).get("sync-on-startup"))
+    def get_sync_interval(self):
+        """
+        Returns the interval in minutes of how often the Argus API is synced if
+        it is None, no re-sync should happen
+        """
+        sync_interval = self.get("api", {}).get("sync-interval", 1)
+        if sync_interval is None:
+            return None
+        return int(sync_interval)
```
What happens if a user configures the value to 0?  A negative value?  A non-integer value?  Sounds like a bad time.

I think `0` should be considered the same as `None` here (i.e. instead have `if not sync_interval:`).  Also, I assume it is meant to otherwise be a positive integer, so if it cannot be coerced to that, this should raise an exception.

lunkwill42 commented:
> But I would argue that we should have a default for syncing (maybe one minute) and if someone does not want re-syncing, they should explicitly set that setting to None

I did not mention a default value, but we also never deploy nav-argus-glue without the sync-on-startup option.  Having `1` minute as default sounds ok to me.  For other considerations, see my  latest review.

lunkwill42 commented on src/navargus/glue.py:
```diff
@@ -559,9 +568,28 @@ def get_api_timeout(self) -> float:
+    def get_sync_interval(self):
+        """
+        Returns the interval in minutes of how often the Argus API is synced if
+        it is None, no re-sync should happen
```
Looks like two sentences were joined together with no period or capitalization.

Referring to [PEP-257](https://peps.python.org/pep-0257/), I suggest this:

```suggestion
        """Returns the configured sync interval in minutes.
        
        The sync interval determines how often a full re-sync against the Argus API
        should take place. A value of `None` means that periodic re-sync should not
        take place.

```

lunkwill42 reviewed (changes requested):
Functionality-wise this looks A-OK.  Just nitpicking docstrings :laughing:

lunkwill42 reviewed (approved):
:tada:
